newIDE/app/src/AssetStore/PrivateAssets/PrivateAssetsAuthorizationProvider.js

```diff
@@ -0,0 +1,151 @@
+import {
+  createAuthorizedUrl,
+  getAuthorizationTokenForPrivateAssets,
+} from '../../Utils/GDevelopServices/Shop';
+import PrivateAssetsAuthorizationContext from './PrivateAssetsAuthorizationContext';
+
+type Props = {| children: React.Node |};
+
+const PrivateAssetsAuthorizationProvider = ({ children }: Props) => {
+  const authenticatedUser = React.useContext(AuthenticatedUserContext);
+  const profile = authenticatedUser.profile;
+  const [authorizationToken, setAuthorizationToken] = React.useState<?string>(
+    null
+  );
+  const isLoading = React.useRef<boolean>(false);
+
+  const fetchAuthorizationToken = React.useCallback(
+    async (userId: string) => {
+      const token = await getAuthorizationTokenForPrivateAssets(
+        authenticatedUser.getAuthorizationHeader,
+        {
+          userId,
+        }
+      );
+      // When a new token is fetched, always update it in the state.
+      setAuthorizationToken(token);
+      return token;
+    },
+    [authenticatedUser.getAuthorizationHeader]
+  );
+
+  const updateAuthorizationToken = async () => {
+    if (!profile) return;
+    const userId = profile.id;
+
+    // If a request is already in progress, don't do anything.
+    if (isLoading.current) return;
+
+    try {
+      isLoading.current = true;
+      await fetchAuthorizationToken(userId);
+    } catch (error) {
+      console.error(error);
+    } finally {
+      isLoading.current = false;
+    }
+  };
+
+  const fetchPrivateAsset = async (
+    assetShortHeader: AssetShortHeader,
+    { environment }: {| environment: Environment |}
+  ): Promise<?Asset> => {
+    if (!profile) return;
+    const userId = profile.id;
+
+    let token =
+      authorizationToken || (await fetchAuthorizationToken(profile.id));
+
+    try {
+      const asset = await getPrivateAsset(assetShortHeader, token, {
+        environment,
+      });
+      return asset;
+    } catch (error) {
+      if (error.response && error.response.status === 404) {
+        // If the token is expired, fetch a new one and try again.
+        token = await fetchAuthorizationToken(userId);
+        const asset = getPrivateAsset(assetShortHeader, token, { environment });
+        return asset;
+      }
+      throw error;
+    }
+  };
+
+  const installPrivateAsset = async ({
+    assetShortHeader,
+    eventsFunctionsExtensionsState,
+    project,
+    objectsContainer,
+    environment,
+  }: InstallAssetShortHeaderArgs): Promise<?InstallAssetOutput> => {
+    if (!profile) {
+      throw new Error(
+        'Unable to install the asset because no profile was found.'
+      );
+    }
+
+    const asset: ?Asset = await fetchPrivateAsset(assetShortHeader, {
+      environment,
+    });
+
+    if (!asset) {
+      throw new Error(
+        'Unable to install the asset because it could not be fetched.'
+      );
+    }
+
+    const token =
+      authorizationToken || (await fetchAuthorizationToken(profile.id));
+
+    const assetWithAuthorizedResourceUrls = {
```

Comment:
This smells weird:
- this object is not typed
- the triple spread indicates something is too complex.
- only a single object asset is handled

A function is a good chance to name what you're doing here. What is the operation being done? Why is this necessary? The name of the function should tell that.